packages/financial-templates-lib/src/price-feed/DefiPulseTotalTVLPriceFeed.js

```diff
@@ -0,0 +1,164 @@
+   * @param {Function} getTime Returns the current time.
+   * @param {Integer} minTimeBetweenUpdates Min number of seconds between updates. If update() is called again before
+   *      this number of seconds has passed, it will be a no-op.
+   */
+  constructor(
+    logger,
+    web3,
+    apiKey,
+    lookback,
+    networker,
+    getTime,
+    minTimeBetweenUpdates,
+    decimals = 18,
+
+  ) {
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+
+    this.apiKey = apiKey;
+
+    this.lookback = lookback;
+    this.networker = networker;
+    this.getTime = getTime;
+    this.minTimeBetweenUpdates = minTimeBetweenUpdates;
+
+    this.toBN = this.web3.utils.toBN;
+
+    this.historicalPrices = [{timestamp:0, value:0}];
+
+    this.convertDecimals = number => {
+      // Converts price result to wei
+      // returns price conversion to correct decimals as a big number
+      return this.toBN(parseFixed(number.toString(), decimals).toString());
+    };
+
+    this.removeValuesOlderThanLookback = () => {
+
+        const currentTime = this.getTime();
+
+        this.historicalPrices = this.historicalPrices.filter(priceObj => priceObj.timestamp > (currentTime - lookback) );
+
+    }
+
+
+  }
+
+  getCurrentPrice() {
+    return this.currentPrice;
+  }
+
+  getHistoricalPrice(time, verbose = false) {
+    if (this.lastUpdateTime === undefined) {
+      return undefined;
+    }
+
+    let closestTime = {timestamp:0, value:0};
+
+    //go through all values and find time that that is the largest and still less than 'time'
+    for(let i = 0; i < this.historicalPrices.length; i++){
+
+      let past = this.historicalPrices[i].timestamp;
+      let val = this.historicalPrices[i].value;
+
+      if( (past > closestTime.timestamp) && (past < time)){
+
+        closestTime.timestamp = past;
+        closestTime.value = val;
+
+      }
+
+    }
+
+    const returnPrice = closestTime.value;
+
+
+    return returnPrice;
+  }
+
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  async update() {
+    const currentTime = this.getTime();
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== undefined && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
+      this.logger.debug({
+        at: "DEFI_PULSE_TOTAL_TVL_PriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTimes + this.minTimeBetweenUpdates - currentTime
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "DEFI_PULSE_TOTAL_TVL_PriceFeed",
+      message: "Updating",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime
+    });
+
+    // 1. Construct URLs.
+    const priceUrl =
+    `https://data-api.defipulse.com/api/v1/defipulse/api/MarketData?api-key=${this.apiKey}`;
```

Comment:
Is there any way to get historical data from defi pulse? It seems as if this class is depending on caching historical data over time internally, but generally the bots expect to be able to immediately access historical data up to the lookback. Does DefiPulse offer any API for immediate historical data?

Reply:
There is a GetHistory endpoint. How should I go about using it? Call it from getCurrentPrice() and store it in state so it can be referenced in getHistoricalPrice(time)?

Reply:
Modifying the CryptoWatchPriceFeed tests now. I looked more into the historical endpoint but the timestamps are a full day apart.  I imagine the bots need more precision than this.  Also, are there preferred ranges for  lookback period and minTimeBetweenUpdates? Defi pulse only updates their api hourly.